engines/twine/debug.cpp

```diff
@@ -0,0 +1,551 @@
+
+	//   - zones window
+	debugAddWindow(205, 55, 634, 97, 4, 0);
+	debugAddWindowText(ZONES_MENU, (int8 *)"You can enable or disable each zone type:");
+	debugAddButton(ZONES_MENU, 205, 118, 350, 138, (int8 *)"Cube Zones", 215, 123, 1, 87, 119, 0, SHOW_ZONE_CUBE);
+	debugAddButton(ZONES_MENU, 205, 139, 350, 159, (int8 *)"Camera Zones", 215, 144, 2, 87, 119, 0, SHOW_ZONE_CAMERA);
+	debugAddButton(ZONES_MENU, 205, 160, 350, 180, (int8 *)"Scenaric Zones", 215, 165, 3, 87, 119, 0, SHOW_ZONE_SCENARIC);
+	debugAddButton(ZONES_MENU, 205, 181, 350, 201, (int8 *)"Celling Grid Zones", 215, 186, 4, 87, 119, 0, SHOW_ZONE_CELLINGGRID);
+	debugAddButton(ZONES_MENU, 205, 202, 350, 222, (int8 *)"Object Zones", 215, 207, 5, 87, 119, 0, SHOW_ZONE_OBJECT);
+	debugAddButton(ZONES_MENU, 205, 223, 350, 243, (int8 *)"Text Zones", 215, 228, 6, 87, 119, 0, SHOW_ZONE_TEXT);
+	debugAddButton(ZONES_MENU, 205, 244, 350, 264, (int8 *)"Ladder Zones", 215, 249, 7, 87, 119, 0, SHOW_ZONE_LADDER);
+}
+
+int32 debugProcessButton(int32 X, int32 Y) {
+	int32 i;
+	int32 j;
+
+	for (i = 0; i < numDebugWindows; i++) {
+		for (j = 0; j < debugWindows[i].numButtons; j++) {
+			if (X > (debugWindows[i].debugButtons[j].left) && X < (debugWindows[i].debugButtons[j].right) && Y > (debugWindows[i].debugButtons[j].top) && Y < (debugWindows[i].debugButtons[j].bottom)) {
+				return (debugWindows[i].debugButtons[j].type);
+			}
+		}
+	}
+
+	return 0;
+}
+
+void debugPlasmaWindow(int8 *text, int32 color) {
+	int32 textSize;
+	processPlasmaEffect(5, color);
+	if (!(rand() % 5)) {
+		plasmaEffectPtr[rand() % 320 * 10 + 6400] = 255;
+	}
+	textSize = getTextSize(text);
+	drawText((SCREEN_WIDTH / 2) - (textSize / 2), 10, text);
+	drawBox(5, 5, 634, 50);
+	copyBlockPhys(5, 5, 634, 50);
+}
+
+void debugProcessWindow() {
+	if (rightMouse) {
+		int32 quit = 0;
+		int8 *text = (int8 *)"Game Debug Window";
+		int32 color = 64;
+		int32 colorIdx = 4;
+		int32 count = 0;
+		MouseStatusStruct mouseData;
+		rightMouse = 0;
+		leftMouse = 0;
+
+		copyScreen(frontVideoBuffer, workVideoBuffer);
+
+		debugResetButtonsState();
+		if (numDebugWindows == 0)
+			debugLeftMenu();
+		debugDrawWindows();
+
+		do {
+			readKeys();
+			getMousePositions(&mouseData);
+
+			if (mouseData.left) {
+				int type = 0;
+				if ((type = debugProcessButton(mouseData.X, mouseData.Y)) != NO_ACTION) { // process menu item
+					if (debugTypeUseMenu(type)) {
+						copyScreen(workVideoBuffer, frontVideoBuffer);
+						copyBlockPhys(205, 55, 634, 474);
+					}
+
+					debugRefreshButtons(type);
+					debugSetActions(type);
+				}
+				mouseData.left = 0;
+			}
+
+			// draw window plasma effect
+			if (count == 256) {
+				colorIdx++;
+				count = 0;
+			}
+			color = colorIdx * 16;
+			if (color >= 240) {
+				color = 64;
+				colorIdx = 4;
+			}
+			debugPlasmaWindow(text, color);
+
+			// quit
+			if (mouseData.right)
+				quit = 1;
+
+			fpsCycles(25); // rest
+
+			count++;
+		} while (!quit);
```

Comment:
Apart from maybe clarifying quit to be a bool rather than an int, something you'll need to do is hunt down every place where you have a loop like this and ensure that you break out of the loop if a call to the engine's shouldQuit() method returns true. Otherwise, the game will be prevented from exiting if the user tries to close the ScummVM window when execution is within the loop

Reply:
True - but I think that I would rather like to convert these ui debug actions to console commands.